crates/benda/src/types/mod.rs

```diff
@@ -0,0 +1,121 @@
+use core::panic;
+
+use bend::fun::Num;
+use bend::imp;
+use pyo3::types::{PyAnyMethods, PyFloat, PyTypeMethods};
+use pyo3::{Bound, FromPyObject, PyAny, PyTypeCheck};
+use tree::{Leaf, Node, Tree};
+
+pub mod f24;
+pub mod i24;
+pub mod tree;
+pub mod u24;
+
+pub trait BendType {
+    fn to_bend(&self) -> imp::Expr;
+}
+
+pub fn extract_inner<'py, T: BendType + PyTypeCheck + FromPyObject<'py>>(
+    arg: Bound<'py, PyAny>,
+) -> Option<T> {
+    let inner = arg.downcast::<T>();
+    if let Ok(inner) = inner {
+        let inner = <T as FromPyObject>::extract_bound(inner.as_any());
+        return Some(inner.unwrap());
+    }
+    None
+}
+
+pub fn extract_num(
+    arg: Bound<PyAny>,
+    t_type: BuiltinType,
+) -> Option<imp::Expr> {
+    match t_type {
+        BuiltinType::I32 => {
+            Some(arg.to_string().parse::<i32>().unwrap().to_bend())
+        }
+        BuiltinType::F32 => {
+            Some(arg.to_string().parse::<f32>().unwrap().to_bend())
+        }
+        _ => unreachable!(),
+    }
+}
+
+pub fn extract_type(arg: Bound<PyAny>) -> Option<imp::Expr> {
+    let t_type = arg.get_type();
+    let name = t_type.name().unwrap();
+
+    let arg_type = BuiltinType::from(name.to_string());
+
+    match arg_type {
+        BuiltinType::U24 => {
+            Some(extract_inner::<crate::u24>(arg).unwrap().to_bend())
+        }
+        BuiltinType::I32 => extract_num(arg, BuiltinType::I32),
+        BuiltinType::F32 => extract_num(arg, BuiltinType::F32),
+        BuiltinType::Tree => {
+            Some(extract_inner::<Tree>(arg).unwrap().to_bend())
+        }
+        BuiltinType::Node => {
+            Some(extract_inner::<Node>(arg).unwrap().to_bend())
+        }
+        BuiltinType::Leaf => {
+            Some(extract_inner::<Leaf>(arg).unwrap().to_bend())
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum BuiltinType {
+    U24,
+    F32,
+    I32,
+    Tree,
+    Leaf,
+    Node,
+}
+
+impl From<String> for BuiltinType {
```

Comment:
Is this being used anywhere?

I don't think we should be relying on strings like this in hte final version.
(if we can rely on in the identity of the types though the Python API and Python itself is not using strings)

Reply:
I'll find a way to reference these types in a better way. I don't like this kind of String usage either.

Reply:
The types names are not mapped anywhere in pyo3. I will find a better solution when our subset are better defined. 
I use this From<String> here:
https://github.com/fglab-tech/benda/blob/28e0bdd1c6b0197441edee2284dc37b0c91589e1/crates/benda/src/types/mod.rs#L48